File: integration/ts/tic-tac-toe.test.ts
/**
 * TicTacToe integration test — stateful contract with two-player game logic.
 *
 * TicTacToe is a StatefulSmartContract with properties:
 *   - playerX: PubKey (readonly, constructor param)
 *   - betAmount: bigint (readonly, constructor param)
 *   - playerO: PubKey (mutable, initialized to 33 zero bytes)
 *   - c0-c8: bigint (mutable, board cells, initialized to 0)
 *   - turn: bigint (mutable, initialized to 0)
 *   - status: bigint (mutable, initialized to 0)
 *   - p2pkhPrefix, p2pkhSuffix: ByteString (readonly, initialized)
 *
 * State-mutating methods: join(opponentPK, sig), move(position, player, sig)
 * Terminal methods: moveAndWin, moveAndTie, cancelBeforeJoin, cancel
 *
 * The SDK auto-computes Sig when null is passed. For multi-player tests,
 * each player needs their own funded wallet and signer. The signer used
 * for a call determines whose private key signs the transaction — the
 * contract's checkSig verifies the signature against the provided pubkey.
 */

import { describe, it, expect } from 'vitest';
import { compileContract } from './helpers/compile.js';
import { RunarContract } from 'runar-sdk';
import { createFundedWallet, createWallet } from './helpers/wallet.js';
import { createProvider } from './helpers/node.js';

describe('TicTacToe', () => {
  it('should compile the TicTacToe contract', () => {
    const artifact = compileContract('examples/ts/tic-tac-toe/TicTacToe.runar.ts');
    expect(artifact).toBeTruthy();
    expect(artifact.contractName).toBe('TicTacToe');
  });

  it('should deploy with playerX and betAmount', async () => {
    const artifact = compileContract('examples/ts/tic-tac-toe/TicTacToe.runar.ts');

    const provider = createProvider();
    const playerX = await createFundedWallet(provider);

    const contract = new RunarContract(artifact, [playerX.pubKeyHex, 5000n]);

    const { txid: deployTxid } = await contract.deploy(provider, playerX.signer, {});
    expect(deployTxid).toBeTruthy();
    expect(typeof deployTxid).toBe('string');
    expect(deployTxid.length).toBe(64);
  });

  it('should join the game as player O (auto-computed state)', async () => {
    const artifact = compileContract('examples/ts/tic-tac-toe/TicTacToe.runar.ts');

    const provider = createProvider();
    const playerX = await createFundedWallet(provider);
    const playerO = await createFundedWallet(provider);

    const contract = new RunarContract(artifact, [playerX.pubKeyHex, 5000n]);
    await contract.deploy(provider, playerX.signer, {});

    // join(opponentPK, sig) — playerO signs, passing their pubkey
    const { txid: joinTxid } = await contract.call(
      'join', [playerO.pubKeyHex, null], provider, playerO.signer,
    );
    expect(joinTxid).toBeTruthy();
    expect(joinTxid.length).toBe(64);

    // After join: status=1 (playing), turn=1 (playerX's turn)
    expect(contract.state.status).toBe(1n);
    expect(contract.state.turn).toBe(1n);
  });

  it('should make a move after join (auto-computed state)', async () => {
    const artifact = compileContract('examples/ts/tic-tac-toe/TicTacToe.runar.ts');

    const provider = createProvider();
    const playerX = await createFundedWallet(provider);
    const playerO = await createFundedWallet(provider);

    const contract = new RunarContract(artifact, [playerX.pubKeyHex, 5000n]);
    await contract.deploy(provider, playerX.signer, {});

    // Player O joins
    await contract.call('join', [playerO.pubKeyHex, null], provider, playerO.signer);

    // Player X moves to position 4 (center) — turn=1 means it's X's turn
    // DEBUG: try manual newState to isolate the issue
    const { txid: moveTxid } = await contract.call(
      'move', [4n, playerX.pubKeyHex, null], provider, playerX.signer,
      { newState: { c4: 1n, turn: 2n } },
    );
    expect(moveTxid).toBeTruthy();
    expect(moveTxid.length).toBe(64);

    // After move: c4=1 (X mark), turn=2 (O's turn)
    expect(contract.state.c4).toBe(1n);
    expect(contract.state.turn).toBe(2n);
  });

  it('should play a full game: X wins top row and claims the pot', async () => {
    const artifact = compileContract('examples/ts/tic-tac-toe/TicTacToe.runar.ts');

    const provider = createProvider();
    const playerX = await createFundedWallet(provider);
    const playerO = await createFundedWallet(provider);

    const betAmount = 1000n;
    const contract = new RunarContract(artifact, [playerX.pubKeyHex, betAmount]);
    await contract.deploy(provider, playerX.signer, { satoshis: Number(betAmount) });

    // Player O joins — doubling the pot (betAmount * 2)
    await contract.call('join', [playerO.pubKeyHex, null], provider, playerO.signer,
      { satoshis: Number(betAmount) * 2 });
    expect(contract.state.status).toBe(1n);

    // X@0, O@3, X@1, O@4 — set up X to win with position 2 (top row)
    await contract.call('move', [0n, playerX.pubKeyHex, null], provider, playerX.signer);
    await contract.call('move', [3n, playerO.pubKeyHex, null], provider, playerO.signer);
    await contract.call('move', [1n, playerX.pubKeyHex, null], provider, playerX.signer);
    await contract.call('move', [4n, playerO.pubKeyHex, null], provider, playerO.signer);

    // Board: X X _ | O O _ | _ _ _  — X plays position 2 to win top row
    expect(contract.state.c0).toBe(1n);
    expect(contract.state.c1).toBe(1n);
    expect(contract.state.turn).toBe(1n); // X's turn

    // moveAndWin(position, player, sig, changePKH, changeAmount)
    // Terminal method — needs terminalOutputs matching the payout the contract expects
    const totalPayout = Number(betAmount) * 2;
    const winnerP2PKH = '76a914' + playerX.pubKeyHash + '88ac';

    const { txid: winTxid } = await contract.call(
      'moveAndWin',
      [2n, playerX.pubKeyHex, null, '00'.repeat(20), 0n],
      provider, playerX.signer,
      {
        terminalOutputs: [
          { scriptHex: winnerP2PKH, satoshis: totalPayout },
        ],
      },
    );
    expect(winTxid).toBeTruthy();
    expect(winTxid.length).toBe(64);
  });

  it('should reject move by wrong player', async () => {
    const artifact = compileContract('examples/ts/tic-tac-toe/TicTacToe.runar.ts');

    const provider = createProvider();
    const playerX = await createFundedWallet(provider);
    const playerO = await createFundedWallet(provider);

    const contract = new RunarContract(artifact, [playerX.pubKeyHex, 5000n]);
    await contract.deploy(provider, playerX.signer, {});

    // Player O joins — turn=1 (X's turn)
    await contract.call('join', [playerO.pubKeyHex, null], provider, playerO.signer);
    expect(contract.state.turn).toBe(1n);

    // Player O tries to move when it's X's turn
    // checkSig(sig, playerO.pubKey) passes, but assertCorrectPlayer fails
    // because turn=1 expects playerX's pubkey
    await expect(
      contract.call('move', [4n, playerO.pubKeyHex, null], provider, playerO.signer),
    ).rejects.toThrow();
  });

  it('should reject join when game is already playing', async () => {
    const artifact = compileContract('examples/ts/tic-tac-toe/TicTacToe.runar.ts');

    const provider = createProvider();
    const playerX = await createFundedWallet(provider);
    const playerO = await createFundedWallet(provider);

    const contract = new RunarContract(artifact, [playerX.pubKeyHex, 5000n]);
    await contract.deploy(provider, playerX.signer, {});

    // Player O joins — status becomes 1
    await contract.call('join', [playerO.pubKeyHex, null], provider, playerO.signer);
    expect(contract.state.status).toBe(1n);

    // Second join attempt — assert(status == 0) fails on-chain
    const anotherPlayer = await createFundedWallet(provider);
    await expect(
      contract.call('join', [anotherPlayer.pubKeyHex, null], provider, anotherPlayer.signer),
    ).rejects.toThrow();
  });
});

File: integration/zig/src/tic_tac_toe_test.zig
const std = @import("std");
const runar = @import("runar");
const helpers = @import("helpers.zig");
const compile = @import("compile.zig");

test "TicTacToe_Compile" {
    const allocator = std.testing.allocator;

    var artifact = compile.compileContract(allocator, "examples/zig/tic-tac-toe/TicTacToe.runar.zig") catch |err| {
        std.log.warn("Could not compile TicTacToe contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    try std.testing.expectEqualStrings("TicTacToe", artifact.contract_name);
    try std.testing.expect(artifact.isStateful());
    std.log.info("TicTacToe compiled: {d} bytes", .{artifact.script.len / 2});
}

test "TicTacToe_Deploy" {
    const allocator = std.testing.allocator;

    if (!helpers.isNodeAvailable(allocator)) {
        std.log.warn("Regtest node not available, skipping test", .{});
        return;
    }

    var artifact = compile.compileContract(allocator, "examples/zig/tic-tac-toe/TicTacToe.runar.zig") catch |err| {
        std.log.warn("Could not compile TicTacToe contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    var player_x = try helpers.newWallet(allocator);
    defer player_x.deinit();

    const px_hex = try player_x.pubKeyHex(allocator);
    defer allocator.free(px_hex);

    const bet_amount: i64 = 5000;

    // Constructor: playerX (PubKey), betAmount (bigint)
    var contract = try runar.RunarContract.init(allocator, &artifact, &[_]runar.StateValue{
        .{ .bytes = px_hex },
        .{ .int = bet_amount },
    });
    defer contract.deinit();

    const fund_txid = try helpers.fundWallet(allocator, &player_x, 1.0);
    defer allocator.free(fund_txid);

    var rpc_provider = helpers.RPCProvider.init(allocator);
    var local_signer = try player_x.localSigner();

    const deploy_txid = try contract.deploy(rpc_provider.provider(), local_signer.signer(), .{ .satoshis = bet_amount });
    defer allocator.free(deploy_txid);

    const utxo = contract.getCurrentUtxo();
    try std.testing.expect(utxo != null);
    try std.testing.expectEqual(@as(usize, 64), deploy_txid.len);
    std.log.info("TicTacToe deployed with betAmount={d}: {s}", .{ bet_amount, deploy_txid });
}

test "TicTacToe_Join" {
    const allocator = std.testing.allocator;

    if (!helpers.isNodeAvailable(allocator)) {
        std.log.warn("Regtest node not available, skipping test", .{});
        return;
    }

    var artifact = compile.compileContract(allocator, "examples/zig/tic-tac-toe/TicTacToe.runar.zig") catch |err| {
        std.log.warn("Could not compile TicTacToe contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    var player_x = try helpers.newWallet(allocator);
    defer player_x.deinit();
    var player_o = try helpers.newWallet(allocator);
    defer player_o.deinit();

    const px_hex = try player_x.pubKeyHex(allocator);
    defer allocator.free(px_hex);
    const po_hex = try player_o.pubKeyHex(allocator);
    defer allocator.free(po_hex);

    const bet_amount: i64 = 5000;

    var contract = try runar.RunarContract.init(allocator, &artifact, &[_]runar.StateValue{
        .{ .bytes = px_hex },
        .{ .int = bet_amount },
    });
    defer contract.deinit();

    const fund_x = try helpers.fundWallet(allocator, &player_x, 1.0);
    defer allocator.free(fund_x);

    var rpc_provider = helpers.RPCProvider.init(allocator);
    var signer_x = try player_x.localSigner();

    const deploy_txid = try contract.deploy(rpc_provider.provider(), signer_x.signer(), .{ .satoshis = bet_amount });
    defer allocator.free(deploy_txid);
    std.log.info("TicTacToe deployed: {s}", .{deploy_txid});

    // Fund playerO
    const fund_o = try helpers.fundWallet(allocator, &player_o, 1.0);
    defer allocator.free(fund_o);
    var signer_o = try player_o.localSigner();

    // join(opponentPK, sig) -- playerO joins
    const join_txid = try contract.call(
        "join",
        &[_]runar.StateValue{
            .{ .bytes = po_hex },
            .{ .int = 0 }, // sig: auto-sign
        },
        rpc_provider.provider(),
        signer_o.signer(),
        null,
    );
    defer allocator.free(join_txid);

    std.log.info("TicTacToe join TX: {s}", .{join_txid});
    try std.testing.expectEqual(@as(usize, 64), join_txid.len);
}

test "TicTacToe_Move" {
    const allocator = std.testing.allocator;

    if (!helpers.isNodeAvailable(allocator)) {
        std.log.warn("Regtest node not available, skipping test", .{});
        return;
    }

    var artifact = compile.compileContract(allocator, "examples/zig/tic-tac-toe/TicTacToe.runar.zig") catch |err| {
        std.log.warn("Could not compile TicTacToe contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    var player_x = try helpers.newWallet(allocator);
    defer player_x.deinit();
    var player_o = try helpers.newWallet(allocator);
    defer player_o.deinit();

    const px_hex = try player_x.pubKeyHex(allocator);
    defer allocator.free(px_hex);
    const po_hex = try player_o.pubKeyHex(allocator);
    defer allocator.free(po_hex);

    const bet_amount: i64 = 5000;

    var contract = try runar.RunarContract.init(allocator, &artifact, &[_]runar.StateValue{
        .{ .bytes = px_hex },
        .{ .int = bet_amount },
    });
    defer contract.deinit();

    const fund_x = try helpers.fundWallet(allocator, &player_x, 1.0);
    defer allocator.free(fund_x);

    var rpc_provider = helpers.RPCProvider.init(allocator);
    var signer_x = try player_x.localSigner();

    const deploy_txid = try contract.deploy(rpc_provider.provider(), signer_x.signer(), .{ .satoshis = bet_amount });
    defer allocator.free(deploy_txid);

    // Fund playerO
    const fund_o = try helpers.fundWallet(allocator, &player_o, 1.0);
    defer allocator.free(fund_o);
    var signer_o = try player_o.localSigner();

    // Join
    const join_txid = try contract.call(
        "join",
        &[_]runar.StateValue{
            .{ .bytes = po_hex },
            .{ .int = 0 },
        },
        rpc_provider.provider(),
        signer_o.signer(),
        null,
    );
    defer allocator.free(join_txid);

    // Move: player X plays position 4 (center)
    const move_txid = try contract.call(
        "move",
        &[_]runar.StateValue{
            .{ .int = 4 }, // position
            .{ .bytes = px_hex }, // player
            .{ .int = 0 }, // sig: auto-sign
        },
        rpc_provider.provider(),
        signer_x.signer(),
        null,
    );
    defer allocator.free(move_txid);

    std.log.info("TicTacToe move TX: {s}", .{move_txid});
    try std.testing.expectEqual(@as(usize, 64), move_txid.len);
}

test "TicTacToe_StateFields" {
    const allocator = std.testing.allocator;

    var artifact = compile.compileContract(allocator, "examples/zig/tic-tac-toe/TicTacToe.runar.zig") catch |err| {
        std.log.warn("Could not compile TicTacToe contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    // TicTacToe should have multiple state fields (board, turn, status, etc.)
    try std.testing.expect(artifact.state_fields.len >= 3);

    // Should have public methods: join, move, moveAndWin
    var public_count: usize = 0;
    for (artifact.abi.methods) |m| {
        if (m.is_public) public_count += 1;
    }
    try std.testing.expect(public_count >= 2);
}
